### PR TITLE
SNOW-779124 add length parameter to StringType

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-779124

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Added the option `length` in `StringType` to specify the maximum number of characters to store.

### CODE REVIEW

**Review by sfc-gh-sfan — COMMENTED** [review]
I wonder how much we could replace to use `StringType._MAX_LENGTH` instead of `16777216`.

File: src/snowflake/snowpark/_internal/type_utils.py [hunk 1]
```diff
@@ -92,6 +92,8 @@ def convert_sf_to_sp_type(
     if column_type_name == "BINARY":
         return BinaryType()
     if column_type_name == "TEXT":
+        if internal_size > 0 and internal_size <= StringType._MAX_LENGTH:
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: I tested it out on server side - `Error: Invalid character length: 0. Must be between 1 and 16,777,216. (line 4)`

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 2]
```diff
@@ -1297,6 +1298,14 @@ def number(precision: int = 38, scale: int = 0) -> str:
         + RIGHT_PARENTHESIS
     )
 
+def string(length: int = 16777216) -> str:
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:1301]: We should do that in case `_MAX_LENGTH` value changes in the future

File: src/snowflake/snowpark/types.py [hunk 3]
```diff
@@ -68,7 +68,13 @@ class DateType(_AtomicType):
 class StringType(_AtomicType):
     """String data type. This maps to the VARCHAR data type in Snowflake."""
 
-    pass
+    _MAX_LENGTH = 16777216
+
+    def __init__(self, length = 16777216) -> None:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/types.py:73]: How about using `None` as the default value?

File: tests/integ/scala/test_dataframe_suite.py [hunk 4]
```diff
@@ -1615,8 +1615,8 @@ def test_createDataFrame_with_given_schema_array_map_variant(session):
     df = session.create_dataframe(data, schema)
     assert (
         str(df.schema)
-        == "StructType([StructField('ARRAY', ArrayType(StringType()), nullable=True), "
-        "StructField('MAP', MapType(StringType(), StringType()), nullable=True), "
+        == "StructType([StructField('ARRAY', ArrayType(StringType(16777216)), nullable=True), "
+        "StructField('MAP', MapType(StringType(16777216), StringType(16777216)), nullable=True), "
```
Comment (sfc-gh-yixie) [inline tests/integ/scala/test_dataframe_suite.py:1619]: `StringType(16777216)` looks a breaking change. By default we may just use `StringType()` then it won't be a breaking change.

File: src/snowflake/snowpark/_internal/type_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: I wonder if we should fail if the size is not within the expected range when initializing `StringType` 🤔

[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: I'm following the logic followed by `DecimalType` a few lines below, which auto adjusts the input `precision` and `scale` to sensible ranges.

Right now, this function is called by values we get from result metadata. I'm not sure how reliable it is but if length value = 0 (which is possible in case when string length is not explicitly defined), then we should treat it as `StringType()` instead of raising error. 

I could possible raise error when length is negative value or great than `MAX_LENGTH`.

[the remark this answers is not in the document]
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: Yeah negative value makes sense. I'm not particularly sure about `MAX_LENGTH` because this might change and is not controlled by client.

File: tests/integ/scala/test_literal_suite.py [hunk 5]
```diff
@@ -30,8 +30,8 @@ def test_literal_basic_types(session):
     assert (
         field_str == "[StructField('ID', LongType(), nullable=False), "
         "StructField('NULL', StringType(), nullable=True), "
-        "StructField('STR', StringType(), nullable=False), "
-        "StructField('CHAR', StringType(), nullable=False), "
+        "StructField('STR', StringType(6), nullable=False), "
+        "StructField('CHAR', StringType(1), nullable=False), "
```
Comment (sfc-gh-aalam) [inline tests/integ/scala/test_literal_suite.py:34]: these lengths come from `result_meta.internal_size`

File: tests/integ/scala/test_dataframe_reader_suite.py [hunk 6]
```diff
@@ -554,13 +554,13 @@ def test_read_parquet_all_data_types_with_no_schema(session, mode):
         StructField('"N"', DecimalType(38, 6), nullable=True),
         StructField('"F"', DoubleType(), nullable=True),
         StructField('"I"', LongType(), nullable=True),
-        StructField('"S"', StringType(), nullable=True),
-        StructField('"C"', StringType(), nullable=True),
+        StructField('"S"', StringType(StringType._MAX_LENGTH), nullable=True),
+        StructField('"C"', StringType(StringType._MAX_LENGTH), nullable=True),
```
Comment (sfc-gh-aalam) [inline tests/integ/scala/test_dataframe_reader_suite.py:558]: the schema is inferred from the size of column stored on server side after parquet file is uploaded. In this case, since size is not specified, it is set to `StringType._MAX_LENGTH`. On the client `StringType() != StringType(StringType._MAX_LENGTH)`

File: src/snowflake/snowpark/_internal/analyzer/schema_utils.py [hunk 7]
```diff
@@ -16,43 +16,43 @@
 
 
 def command_attributes() -> List[Attribute]:
-    return [Attribute('"status"', StringType())]
+    return [Attribute('"status"', StringType(StringType._MAX_LENGTH))]
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/schema_utils.py:19]: On server, these columns are not set with `STRING(LENGTH)` so the default value it takes is `STRING(16777216)`. On the client we have to explicitly set it to `StringType(_MAX_LENGTH)` because here `StringType() != StringType(_MAX_LENGTH)`.

File: src/snowflake/snowpark/types.py (hunk 3 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/types.py]: I am using `None` as default value but it creates a difference between client and server in the meaning of uninitialized string type. On server `STRING = STRING(16777216)` but on client `StringType() != StringType(16777216)`

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 8]
```diff
@@ -97,6 +97,7 @@
 ROW_COUNT = "ROWCOUNT"
 RIGHT_ARROW = " => "
 NUMBER = " NUMBER "
+STRING = "STRING"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:100]:
```suggestion
STRING = " STRING "
```

File: src/snowflake/snowpark/types.py [hunk 9]
```diff
@@ -68,7 +68,15 @@ class DateType(_AtomicType):
 class StringType(_AtomicType):
     """String data type. This maps to the VARCHAR data type in Snowflake."""
 
-    pass
+    _MAX_LENGTH = 16777216
+
+    def __init__(self, length = None) -> None:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/types.py:73]:
```suggestion
    def __init__(self, length: Optional[int] = None) -> None:
```

File: src/snowflake/snowpark/_internal/type_utils.py [hunk 10]
```diff
@@ -92,7 +92,11 @@ def convert_sf_to_sp_type(
     if column_type_name == "BINARY":
         return BinaryType()
     if column_type_name == "TEXT":
-        return StringType()
+        if internal_size > 0:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/type_utils.py:95]: Do you know that this internal_size is usually 0 or a positive number?

File: tests/integ/scala/test_dataframe_reader_suite.py (hunk 6 above)
Comment (sfc-gh-jdu) [reply tests/integ/scala/test_dataframe_reader_suite.py:558]: Can't we make it equivalent by overwriting `__eq__`?

File: src/snowflake/snowpark/_internal/type_utils.py (hunk 10 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: A valid column that is already created in the server will be positive number. If we try to create a number that is negative or 0, server will throw an error

File: src/snowflake/snowpark/dataframe_reader.py [hunk 11]
```diff
@@ -482,14 +482,21 @@ def _read_semi_structured_file(self, path: str, format: str) -> DataFrame:
...
-                        precision = int(data_type_parts[1].split(",")[0])
-                        scale = int(data_type_parts[1].split(",")[1][:-1])
+                        if "," in data_type_parts[1]:
+                            precision = int(data_type_parts[1].split(",")[0])
+                            scale = int(data_type_parts[1].split(",")[1][:-1])
+                            internal_size = 0
+                        else:
+                            precision = 0
+                            scale = 0
+                            internal_size = int(data_type_parts[1][:-1])
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe_reader.py:495]: Curious: how is this related to the `StringType` change? 

Also, test coverage.

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/dataframe_reader.py (hunk 11 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe_reader.py:495]: The size of a string column comes in result metadata inside the `internal_size` field. So we use this to create `StringType()` on the client

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe_reader.py:495]: I couldn't find a way to specify string column max length in dataframe reader format so I'll remove this section and default column size to 0

File: src/snowflake/snowpark/_internal/type_utils.py (hunk 10 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: Then should we raise an exception when it's negative or 0?

File: src/snowflake/snowpark/types.py [hunk 12]
```diff
@@ -68,7 +68,33 @@ class DateType(_AtomicType):
 class StringType(_AtomicType):
     """String data type. This maps to the VARCHAR data type in Snowflake."""
 
-    pass
+    _MAX_LENGTH = 16777216
+
+    def __init__(self, length : Optional[int] = None) -> None:
+        self.length = length
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/types.py:74]: type hint and docstring since it's a public property now

File: src/snowflake/snowpark/types.py [hunk 13]
```diff
@@ -68,7 +68,33 @@ class DateType(_AtomicType):
...
+
+    def __repr__(self) -> str:
+        if self.length:
+            return f"StringType({self.length})"
+        return "StringType()"
+
+    def __eq__(self, other):
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/types.py:81]: can we add a comment here and explaining when the equality holds?

File: src/snowflake/snowpark/types.py [hunk 14]
```diff
@@ -68,7 +68,33 @@ class DateType(_AtomicType):
...
+        if not isinstance(other, StringType):
+            return False
+
+        if self.length == other.length:
+            return True
+
+        if self.length in (None, StringType._MAX_LENGTH) and other.length in (
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/types.py:88]: nit: (None, StringType._MAX_LENGTH) can be a constant?

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/_internal/type_utils.py (hunk 10 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/type_utils.py:95]: For negative, yes we should raise and error but as of now, we also use 0 to signify that internal size is not declared so it is a proxy for `None`.

File: tests/integ/scala/test_literal_suite.py [hunk 15]
```diff
@@ -27,17 +27,30 @@ def test_literal_basic_types(session):
...
+            "StructField('BYTES', BinaryType(), nullable=False), "
+            "StructField('INT', LongType(), nullable=False), "
+            "StructField('FLOAT', DoubleType(), nullable=False), "
+            "StructField('DECIMAL', DecimalType(38, 18), nullable=False)]"
+        )
+    else:
+        assert (
+            field_str == "[StructField('ID', LongType(), nullable=False), "
+            "StructField('NULL', StringType(16777216), nullable=True), "
+            "StructField('STR', StringType(6), nullable=False), "
```
Comment (sfc-gh-aalam) [inline tests/integ/scala/test_literal_suite.py:46]: we need if-else because we are comparing `str(schema)` instead of `schema`

File: tests/integ/scala/test_literal_suite.py [hunk 16]
```diff
@@ -122,11 +135,18 @@ def test_special_literals(session):
...
+    if session.sql_simplifier_enabled:
+        assert (
+            str(df.schema) == "StructType([StructField('ID', LongType(), nullable=False), "
+            "StructField('NULL', StringType(), nullable=True), "
+            "StructField('LITERAL', LongType(), nullable=False)])"
+        )
+    else:
+        assert (
+            str(df.schema) == "StructType([StructField('ID', LongType(), nullable=False), "
+            "StructField('NULL', StringType(16777216), nullable=True), "
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply tests/integ/scala/test_literal_suite.py:147]: @sfc-gh-yixie can this be considered behavior change? In case when sql simplified is disabled, in this test we compare `str(schema)` instead of `schemas`. With the updated `__eq__` comparing schemas still evaluates to be equal

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 17]
```diff
@@ -1297,6 +1298,16 @@ def number(precision: int = 38, scale: int = 0) -> str:
         + RIGHT_PARENTHESIS
     )
 
+def string(length: Optional[int] = None) -> str:
+    if length:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:1302]: Can length be 0? If so you may want to use `if length is None:` here. If no this this is fine.

File: src/snowflake/snowpark/types.py [hunk 18]
```diff
@@ -66,9 +66,47 @@ class DateType(_AtomicType):
...
 class StringType(_AtomicType):
-    """String data type. This maps to the VARCHAR data type in Snowflake."""
+    """String data type. This maps to the VARCHAR data type in Snowflake.
 
-    pass
+    A ``StringType`` object can be created in the following ways::
+
+        >>> string_t = StringType(23)  # this can be used to create a string type column which holds at most 23 chars
+        >>> string_t = StringType()    # this can be used to create a string type column with maximum allowed length
+        >>> string_t = StringType(StringType._MAX_LENGTH)  # same as creating StringType()
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/types.py:75]: If you expose MAX_LENGTH it should be a public API so no prefix understore

**Review by sfc-gh-yixie — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py (hunk 17 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:1302]: Length can't be zero. Server raises an error if we do so
